corehq/apps/sso/tests/test_backends.py

```diff
@@ -0,0 +1,225 @@
+        AuthenticatedEmailDomain.objects.all().delete()
+        super().tearDown()
+
+    def _set_saml_session_index(self):
+        self.request.session['samlSessionIndex'] = '_7c84c96e-8774-4e64-893c-06f91d285100'
+
+    def test_backend_failure_without_idp_slug(self):
+        """
+        SsoBackend should fail to move past the first check because an idp_slug
+         was not passed to authenticate()
+        """
+        self._set_saml_session_index()
+        user = auth.authenticate(
+            request=self.request,
+            username=self.user.username
+        )
+        self.assertIsNone(user)
+        with self.assertRaises(AttributeError):
+            self.request.sso_login_error
+
+    def test_backend_failure_without_saml_session_index(self):
+        """
+        SsoBackend should fail to move past the first check because a
+         samlSessionIndex is not present in request.session.
+        """
+        user = auth.authenticate(
+            request=self.request,
+            username=self.user.username,
+            idp_slug=self.idp.slug
+        )
+        self.assertIsNone(user)
+
+    def test_login_error_if_idp_doesnt_exist(self):
+        """
+        SsoBackend should fail to return a user if the Identity Provider
+        slug associated with that user does not exist. It should also populate
+        request.sso_login_error with an error message.
+        """
+        self._set_saml_session_index()
+        user = auth.authenticate(
+            request=self.request,
+            username=self.user.username,
+            idp_slug='doesnotexist'
+        )
+        self.assertIsNone(user)
+        self.assertEqual(
+            self.request.sso_login_error,
+            "Identity Provider does not exist."
+        )
+
+    def test_login_error_if_idp_not_active(self):
+        """
+        SsoBackend should fail to return a user if the Identity Provider
+        associated with that user is not active. It should also populate
+        request.sso_login_error with an error message.
+        """
+        self._set_saml_session_index()
+        self.idp.is_active = False
+        self.idp.save()
+        user = auth.authenticate(
+            request=self.request,
+            username=self.user.username,
+            idp_slug=self.idp.slug
+        )
+        self.assertIsNone(user)
+        self.assertEqual(
+            self.request.sso_login_error,
+            "This Identity Provider is not active."
+        )
+
+    def test_login_error_if_bad_username(self):
+        """
+        SsoBackend should fail to return a user if the username passed to it
+        is not a valid email address (missing email domain / no `@`). It should
+        also populate request.sso_login_error with an error message.
+        """
+        self._set_saml_session_index()
+        user = auth.authenticate(
+            request=self.request,
+            username='badusername',
+            idp_slug=self.idp.slug
+        )
```

Comment:
Just a suggestion, but I've liked creating helper methods for situations like this, where only one of the parameters is relevant to the test (`username='badusername'`). Creating sensible defaults for all of the parameters then allows you to just do something like: `self.authenticate(username='badusername')`